utility methods to parse B.C. date strings and early numeric date strings.

ndushay opened the pull request.

cbeer commented on spec/date_parsing_spec.rb:
```diff
@@ -366,6 +359,17 @@
   }
+  # we have data like this for our Roman coins collection
+  early_numeric_dates = {
+    '-999' => '-001',
```
Are these mappings right? I don't understand.

ndushay replied:
good suggestion.  done.

cbeer commented on lib/stanford-mods/date_parsing.rb:
```diff
@@ -85,6 +86,55 @@ def self.facet_string_for_century(date_str)
+      end
+
+      # get String sortable value from date String containing yyy, yy, y, -y, -yy, -yyy
+      #  note that these values must *lexically* sort to create a chronological sort.
```
Is that only true because we're indexing them as strings?

ndushay replied:
a-yup.  But changing SW isn't in scope, I don't thing, so we're gonna leave it there.

ndushay commented:
If we want to have an integer sort field, i can add methods for it, and eventually maybe SW will use it.  Not sure I should do that now - seems like getting feigenbaum working is more important.

jkeck commented on lib/stanford-mods/date_parsing.rb:
```diff
@@ -85,6 +86,55 @@ def self.facet_string_for_century(date_str)
+      # @param [String] date_str String containing B.C.
+      # @return [String, nil] ddd B.C.  if ddd B.C. in pattern; nil otherwise
+      def self.facet_string_for_bc(date_str)
+        bc_matches = date_str.match(/\d{1,4}.*B\.C\./) if date_str
```
This seems to be a duplicate of the logic above.  Maybe this can be abstracted out into a private method that both methods can take advantage of.

ndushay replied:
I thought about this, but it didn't seem like it would make the code clearer to abstract it.  If it was more than one line, I would be more likely to do so.  What do you think?  Would it actually make it easier or harder to follow what is going on in the code?

jkeck replied:
IMO having a method with a name that describes what it is we're checking is clearer than having to mentally parse the regex (or realize it's the same regex as above once I've read half of it).  At least then, once I have mentally come to terms with what it is that method checks, I don't have to re-parse when I see a reference to that method again.  (I may just be a DRY stickler though)

FWIW, that makes a lot more sense if we are using an instance of the class so we could be more declarative like just saying `return unless bc_date_match?` or something like that.

ndushay replied:
ok, will change regex's to methods.  personally, I get tired of jumping around in the code to read one liner methods scattered up and down throughout a class, but good naming can help with that.  Happy to oblige you, Obi wan!

jkeck commented on lib/stanford-mods/date_parsing.rb:
```diff
@@ -85,6 +86,55 @@ def self.facet_string_for_century(date_str)
+      #    -700 for 300 B.C., -750 for 250 B.C., -800 for 200 B.C., -801 for 199 B.C.
+      # @param [String] date_str String containing B.C.
+      # @return [String, nil] String sortable -ddd if B.C. in pattern; nil otherwise
+      def self.sortable_year_for_bc(date_str)
```
I know this pattern existed before this commit, but something is jumping out at me.

Seeing that this is a class that defines all its methods as class methods (`def self.whatever`), then passes the same variable between all of the methods, seems odd to me.  This either feels like a collection of methods that another class would include or possibly that there should be an initialize method that can take the `date_str` as and argument and then all of these methods could be messages to send to the `DateParsing` class.

```ruby
date_parser = Stanford::Mods::DateParsing.new(my_date_string)

date_parser.sortable_year_for_bc
date_parser.facet_string_for_bc
```

That may be totally out of scope for this PR, but thought I would just give my $0.02 on that.

ndushay replied:
there's a PR coming with a method that ties all of these together.  The reason they are all class methods is so that other code (outside stanford-mods) can use these methods, once this is split off into a gem.

ndushay replied:
I hear ya - wondering if it is less overhead / clearer to downstream class to instantiate a DateParsing object   vs.   utilizing scoped methods like 

- Stanford::Mods::DateParsing.facet_string_from_date_str(my_date_string)
- Stanford::Mods::DateParsing.sortable_string_from_date_str(my_date_string)

I may be thinking too much about java practices and solrmarc crap ... but the notion of class methods, such as there are for Date and DateTime was what I was thinking about, TBH.

jkeck replied:
If that's the API that we're trying to accomplish (and we really never intend on instantiating an instance of `DateParsing`) then I wonder if it makes more sense for this to be a module.

And also, I should mention, this was just a my $0.02 kinda thing.  It really doesn't matter ultimately.  Classes are just a special version of modules, but I think of them more specifically as things that I will create instances of.

jkeck commented on lib/stanford-mods/date_parsing.rb:
```diff
@@ -85,6 +86,55 @@ def self.facet_string_for_century(date_str)
+      #   positive number strings will be left padded with zeros for clarity in the facet
+      # @param [String] date_str String containing yyy, yy, y, -y, -yy, -yyy
+      def self.facet_string_for_early_numeric(date_str)
+        return unless date_str.match(/^\-?\d{1,3}$/)
```
Same thing w/ this matching expression, I wonder if this could be moved into a method that both methods that repeat this code could use instead (and also may be more expressive).

ndushay replied:
ditto with above.  Jumping down to the private mehtods to find out what the regex is seemed less desirable to me, but I am malleable on this point.

ndushay commented:
I know this is all weird with "wait, wait, it's coming!" sort of comments, but date parsing is tricky and I sort of started from the inside out.   I do have plans to split the actual date string parsing into a gem (see TODO comment at top of class) and I also have code ready that utilizes all these methods to do a really good job of parsing a wide variety of date strings.   They all take a string param, it's true ... but that seems most realistic for most of our code that will need to parse dates.

ndushay commented:
Was just about to move the regex-es into private methods and then got a "wait, wait - all the other regex are right in the methods using them." so I think it makes this code more inconsistent and less readable. 

The comments are certainly welcome, but I think we may be having a bit of a "potayto potahto" moment here ... and I feel the goal of getting this indexing working for Feigenbaum moving further and further away.  How strongly do you feel about your points?   I hear ya, but i don't necessarily agree with you as yet.

I had already addressed the concerns Chris brought up and this PR has been sitting since Friday afternoon.  The refactoring you're suggesting doesn't make this code significantly more readable or maintainable, IMO.  Wanting to instantiate the objects -- that is a great idea ... perhaps when this code is split out into its own gem.  For now, it's a bunch of utility methods that I didn't think belonged in the StanfordMods::Record object.

If you feel strongly, let's talk F2F.  I fear this medium is not my favorite when there are differences of opinions.

jkeck commented:
:+1:

ndushay commented:
merged per Jessie's :+1:  (and in person conversation)
